### PR TITLE
Unify API Span Start/End Options

### LINKED ISSUE #988: Unify API Span Start/End Options
From [recent feedback](https://docs.google.com/document/d/1WQ9_KQTAu2cfkh4A0OZFB9DfEtZkBa-1x10JuQJK5mM/edit):

> We may consider not having both StartOption/EndOption and StartConfig/EndConfig. Having two types for the same purpose makes the packages verbose and confusing.  We had them both in the OpenCensus Go and they didn’t really add too much value and confused the users.

Additionally I have received requests to allow things like the span kind be set when the span is ended. It seems like unifying these into a single option passed to the `Start` and `End` functions for `Spans` would indeed clean up the API and hand the control of configuration to the user (instead of this project deciding on what can be configured at start or end).

### PR DESCRIPTION
**Pull request by MrAlias**
- Replace `StartOption` and `EndOption` in the trace API with `SpanOption`.
- Add a unified `SpanConfig` to match and a `SpanConfigure` function to parse a `SpanConfig` from `SpanOption`s.
- Replace the `func LinkedTo(sc SpanContext, attrs ...label.KeyValue) StartOption` with `func WithLinks(links ...Link) SpanOption`. This is more consistent with our other option patterns and a cleaner function signature.

Resolves #988 
Part of #536

### CODE REVIEW

File: api/trace/api.go [hunk 1]
```diff
@@ -137,18 +121,106 @@ type Span interface {
...
+func WithRecord() SpanOption {
+	return recordSpanOption(true)
+}
+
+type newRootSpanOption bool
+
+func (o newRootSpanOption) Apply(c *SpanConfig) { c.NewRoot = bool(o) }
+
+// WithNewRoot specifies that the Span should be treated as a root Span. The
+// parent span context should be ignored when defining the Span's trace
```
Comment (evantorrie) [inline api/trace/api.go:211]:
```suggestion
// WithNewRoot specifies that the Span should be treated as a root Span. Any
// existing parent span context will be ignored when defining the Span's trace
```

File: api/trace/api.go [hunk 2]
```diff
@@ -137,18 +121,106 @@ type Span interface {
...
+	Timestamp time.Time
+	// Links are the associations a Span has with other Spans.
+	Links []Link
+	// Record is the recording state of a Span.
+	Record bool
+	// NewRoot identifies a Span as the root Span for a new trace. This is
+	// commonly used when an existing trace crosses trust boundaries and the
+	// remote parent span context should be ignored for security.
+	NewRoot bool
+	// SpanKind is the SpanKind a Span is.
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply api/trace/api.go]: Gave it another try. PTAL

File: api/trace/api.go [hunk 3]
```diff
@@ -100,7 +84,7 @@ type Span interface {
 
 	// End completes the span. No updates are allowed to span after it
 	// ends. The only exception is setting status of the span.
-	End(options ...EndOption)
+	End(options ...SpanOption)
```
Comment (jmacd) [inline api/trace/api.go:87]: Should `End()` document which of the SpanOptions can take effect? It looks like only the timestamp.

Comment (MrAlias) [reply api/trace/api.go:87]: I think it should, but only on the implementation side. I added comments for [the default SDK](https://github.com/open-telemetry/opentelemetry-go/pull/1108/files#diff-06fc1a413b40f256294db86b8f7a33b2R111-R117) to specify exactly the implementation details that you mentioned. It seemed incorrect to place restrictions what is supported from the implementation level into the API.

Comment (MrAlias) [reply api/trace/api.go:103]: Going to resolve. Please let me know if this was in error.

**Pull request merged**